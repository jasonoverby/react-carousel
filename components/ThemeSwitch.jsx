import React from 'react';
import PropTypes from 'prop-types';

export const ThemeSwitch = ({ checked, toggleTheme }) => {
  const handleKeyPress = (e) => {
    // if user presses spacebar when switch is in focus
    // theme is toggled
    if (e.keyCode === 32) {
      // prevents default browser scroll on spacebar behavior
      e.preventDefault();
      toggleTheme();
    }
  };
  return (
    <>
      <span className="themeIcon">ðŸŒž</span>
      <div className="themeSwitch">
        <label tabIndex="0" onKeyDown={handleKeyPress}>
          <input
            type="checkbox"
            role="switch"
            onChange={toggleTheme}
            checked={checked}
            aria-checked={checked}
          />
          <span className="slider"></span>
        </label>
      </div>
      <span className="themeIcon">ðŸŒœ</span>
    </>
  );
};

ThemeSwitch.propTypes = {
  checked: PropTypes.bool,
  toggleTheme: PropTypes.func,
};
