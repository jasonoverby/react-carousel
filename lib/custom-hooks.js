import { useEffect } from 'react';

import { getNewImageIndex, updateImage } from './utils';

export const useImageCycling = ({
  carouselPaused,
  imageIndex,
  setActiveDotIndex,
  setImageIndex,
  setImageLoaded,
}) => {
  useEffect(() => {
    const intervalId = setInterval(async () => {
      if (!carouselPaused) {
        const updatedImageIndex = getNewImageIndex(imageIndex, 'increment');
        await updateImage({
          setActiveDotIndex,
          setImageIndex,
          setImageLoaded,
          updatedImageIndex,
        });
      }
    }, 10000);
    return () => clearInterval(intervalId);
  });
};
