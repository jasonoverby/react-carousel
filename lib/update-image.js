import { imageLocations } from './image-locations';
import { wait } from './wait';

const NUM_IMAGES = imageLocations.length;

export const getNewImageIndex = (imageIndex, incrementOrDecrement) => {
  if (incrementOrDecrement === 'increment') {
    return imageIndex === NUM_IMAGES - 1 ? 0 : imageIndex + 1;
  }
  if (incrementOrDecrement === 'decrement') {
    return imageIndex === 0 ? NUM_IMAGES - 1 : imageIndex - 1;
  }

  return imageIndex;
};

export const updateImage = async ({
  updatedImageIndex,
  setImageIndex,
  setImageLoaded,
}) => {
  // hide current image
  setImageLoaded(false);
  // ensure previous image opacity has time to ease to 0
  // when arrow or dot is clicked
  await wait(200);
  setImageIndex(updatedImageIndex);
  // ensure previous image opacity has time to ease to 0
  // when cycling to next image automatically
  await wait(500);
  // show updated image
  setImageLoaded(true);
};
