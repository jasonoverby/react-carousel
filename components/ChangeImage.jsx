import React from 'react';
import PropTypes from 'prop-types';

import { updateImage, getNewImageIndex } from '/lib/utils';

export const DecrementImage = ({
  imageIndex,
  setActiveDotIndex,
  setImageIndex,
  setImageLoaded,
}) => (
  <button
    aria-label="show previous image"
    className="changeImage"
    onClick={async () => {
      const updatedImageIndex = getNewImageIndex(imageIndex, 'decrement');
      await updateImage({
        setActiveDotIndex,
        setImageIndex,
        setImageLoaded,
        updatedImageIndex,
      });
    }}
  >
    <span className="arrow">◀</span>
  </button>
);

export const IncrementImage = ({
  imageIndex,
  setActiveDotIndex,
  setImageIndex,
  setImageLoaded,
}) => (
  <button
    aria-label="show previous image"
    className="changeImage"
    onClick={async () => {
      const updatedImageIndex = getNewImageIndex(imageIndex, 'increment');
      await updateImage({
        setActiveDotIndex,
        setImageIndex,
        setImageLoaded,
        updatedImageIndex,
      });
    }}
  >
    <span className="arrow">▶</span>
  </button>
);

DecrementImage.propTypes = {
  imageIndex: PropTypes.number,
  setActiveDotIndex: PropTypes.func,
  setImageIndex: PropTypes.func,
  setImageLoaded: PropTypes.func,
};
IncrementImage.propTypes = DecrementImage.propTypes;
