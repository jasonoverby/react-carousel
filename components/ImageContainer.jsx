import React from 'react';
import PropTypes from 'prop-types';

import { imageLocations } from '../lib/image-locations';

export const ImageLink = ({ imageIndex, imageLoaded, setCarouselPaused }) => {
  return (
    <a
      aria-label={`Visit
        ${imageLocations[imageIndex].href} (opens in a new tab)`}
      href={imageLocations[imageIndex].href}
      onMouseOver={() => setCarouselPaused(true)}
      // for keyboard navigation
      onFocus={() => setCarouselPaused(true)}
      onMouseOut={() => setCarouselPaused(false)}
      // for keyboard navigation
      onBlur={() => setCarouselPaused(false)}
      rel="noreferrer"
      target="_blank"
    >
      <img
        alt={imageLocations[imageIndex].alt}
        aria-label="This image will change every ten seconds unless the cursor is hovering over it"
        src={imageLocations[imageIndex].src}
        className={imageLoaded ? 'imageLoaded' : ''}
      />
    </a>
  );
};

export const ImageContainer = ({
  imageIndex,
  imageLoaded,
  setCarouselPaused,
}) => (
  <div className="imgContainer">
    <ImageLink
      imageIndex={imageIndex}
      imageLoaded={imageLoaded}
      setCarouselPaused={setCarouselPaused}
    />
  </div>
);

ImageContainer.propTypes = {
  imageIndex: PropTypes.number.isRequired,
  imageLoaded: PropTypes.bool.isRequired,
  setCarouselPaused: PropTypes.func.isRequired,
};
ImageLink.propTypes = ImageContainer.propTypes;
