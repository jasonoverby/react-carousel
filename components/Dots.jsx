import React from 'react';
import PropTypes from 'prop-types';

import { imageLocations } from '../lib/image-locations';
import { updateImage } from '../lib/update-image';

export const Dots = ({ imageIndex, setImageIndex, setImageLoaded }) => {
  const dots = imageLocations.map((_, i) => (
    <Dot
      imageIndex={imageIndex}
      index={i}
      key={`dot-${i}`}
      setImageIndex={setImageIndex}
      setImageLoaded={setImageLoaded}
    />
  ));

  return <div className="dots">{dots}</div>;
};

export const Dot = ({ imageIndex, index, setImageIndex, setImageLoaded }) => {
  return (
    <button
      className={index === imageIndex ? 'active' : ''}
      // image number in label should not be zero-indexed as that might be confusing
      aria-label={`show image ${index + 1}`}
      onClick={async () => {
        await updateImage({
          updatedImageIndex: index,
          setImageIndex,
          setImageLoaded,
        });
      }}
      type="button"
    >
      <span>â€¢</span>
    </button>
  );
};

Dots.propTypes = {
  setImageIndex: PropTypes.func.isRequired,
  setImageLoaded: PropTypes.func.isRequired,
};
Dot.propTypes = {
  ...Dots.propTypes,
  index: PropTypes.number.isRequired,
};
