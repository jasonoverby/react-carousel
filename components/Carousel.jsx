import React, { useState } from 'react';

import { ImageContainer } from './ImageContainer';
import { Dots } from './Dots';
import { DecrementImage, IncrementImage } from './ChangeImage';
import { useImageCycling } from '../lib/custom-hooks';

export const Carousel = () => {
  const [imageIndex, setImageIndex] = useState(0);
  const [carouselPaused, setCarouselPaused] = useState(false);
  const [imageLoaded, setImageLoaded] = useState(true);

  useImageCycling({
    carouselPaused,
    imageIndex,
    setImageIndex,
    setImageLoaded,
  });

  return (
    <div className="container">
      <DecrementImage
        imageIndex={imageIndex}
        setImageIndex={setImageIndex}
        setImageLoaded={setImageLoaded}
      />
      <ImageContainer
        imageIndex={imageIndex}
        imageLoaded={imageLoaded}
        setCarouselPaused={setCarouselPaused}
      />
      <IncrementImage
        imageIndex={imageIndex}
        setImageIndex={setImageIndex}
        setImageLoaded={setImageLoaded}
      />
      <Dots
        imageIndex={imageIndex}
        setImageIndex={setImageIndex}
        setImageLoaded={setImageLoaded}
      />
    </div>
  );
};
