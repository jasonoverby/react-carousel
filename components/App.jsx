import React from 'react';

import { Carousel } from './Carousel';

export const App = () => {
  return <Carousel />;
};
