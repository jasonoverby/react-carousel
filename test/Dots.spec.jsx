import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import * as updateImage from '../lib/update-image';
import { Dots, Dot } from '../components/Dots';

jest.spyOn(updateImage, 'updateImage');

describe('Dots and Dot', () => {
  const setImageIndex = jest.fn();
  const setImageLoaded = jest.fn();

  describe('Dots', () => {
    it('renders correctly', () => {
      const { asFragment } = render(
        <Dots setImageIndex={setImageIndex} setImageLoaded={setImageLoaded} />,
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('Dot', () => {
    const index = 0;
    it('renders correctly', () => {
      const { asFragment } = render(
        <Dot
          index={index}
          setImageIndex={setImageIndex}
          setImageLoaded={setImageLoaded}
        />,
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it('calls updateImage on click', () => {
      const { container } = render(
        <Dot
          index={index}
          setImageIndex={setImageIndex}
          setImageLoaded={setImageLoaded}
        />,
      );

      const button = container.firstChild;
      fireEvent.click(button);
      expect(updateImage.updateImage).toHaveBeenCalledWith({
        updatedImageIndex: index,
        setImageIndex,
        setImageLoaded,
      });
    });
  });
});
