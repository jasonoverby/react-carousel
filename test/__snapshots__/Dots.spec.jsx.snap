// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dots and Dot Dot renders correctly 1`] = `
<DocumentFragment>
  <button
    aria-label="show image 1"
    class=""
    type="button"
  >
    <span>
      •
    </span>
  </button>
</DocumentFragment>
`;

exports[`Dots and Dot Dots renders correctly 1`] = `
<DocumentFragment>
  <div
    class="dots"
  >
    <button
      aria-label="show image 1"
      class=""
      type="button"
    >
      <span>
        •
      </span>
    </button>
    <button
      aria-label="show image 2"
      class=""
      type="button"
    >
      <span>
        •
      </span>
    </button>
    <button
      aria-label="show image 3"
      class=""
      type="button"
    >
      <span>
        •
      </span>
    </button>
  </div>
</DocumentFragment>
`;
